### PR TITLE
docs: add mutex improvement proposal

### PR DESCRIPTION
**Pull request by isubasinghe**
Related to #8684

### CODE REVIEW

**Comment by sarabala1979** [comment]
@isubasinghe Can you present in the contributor meeting? We can discuss

**Review by caelan-io — APPROVED** [review]

**Comment by isubasinghe** [comment]
update: Discussed with @sarabala1979 what the next steps are for this. Bala mentioned to try replace the getResourceKey with getHolderKey. I thought during the meeting this would work as well, looking into that again this will not work without changing some data in one of the structs. 

@sarabala1979 could you also please clarify here why we can't go with the config map approach? I believe it was about push back regarding introducing another resource? 

The next steps for this would be to go with modifying the format of the key in the MutexHolding struct. @sarabala1979 with your okay, I will proceed with this fix.

**Comment by isubasinghe** [comment]
not stale and not resolved.

**Comment by isubasinghe** [comment]
Hey @juliev0 or @terrytangyuan  could you please give me some feedback on this if you have any time available? Mutexes have been broken since they were implemented. I have a partial solution here: https://github.com/argoproj/argo-workflows/pull/10009.
 
But I need to know if I can go ahead with this approach before continuing to invest more time/money in it. 

TL;DR -> store a couple more mutex related information in config maps

**Comment by terrytangyuan** [comment]
I haven't got a chance to look into details yet but would this be a breaking change for current users who rely on mutexes?

**Comment by isubasinghe** [comment]
@terrytangyuan it shouldn't change anything for users, it will introduce an additional config map where mutex related data is stored.

File: docs/proposals/mutex-improvements.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+* Relatively simple change.
+* Single source of truth
+
+##### Disadvantages of solution #1
+
+* Changing a field could be major change and might break projects in the ecosystem that rely on this behavior. This has been the behavior of mutexes ever since they have been implemented, it is high risk to change this in my opinion.
+
+#### Store holding/pending information in a config map
+
+We can store the holder keys inside a config map, on release, we refer to the holder keys inside this config map. There is an existing WIP PR for this, [here](https://github.com/argoproj/argo-workflows/pull/10009).
```
Comment (sarabala1979) [inline docs/proposals/mutex-improvements.md:71]: Do you think there is a security concern in multi-tenant env?  If Controller updates the holder on Configmap, The controller should have cluster-level RBAC to update the config map across the namespace.

File: docs/proposals/mutex-improvements.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+### How should they work?
+
+We should not be generating information after it has already been generated. It is extremely important to only maintain a single source of truth,
+generating this information (the holder key) produces sources of truth at every generation point. Ideally we should be storing this information somewhere
+as we will be needing it eventually when a release call is made.
+
+### Solutions
+
+#### Store the holder key directly in the MutexStatus/SemaphoreStatus structs
```
Comment (sarabala1979) [inline docs/proposals/mutex-improvements.md:53]: How will these solutions form the holder keys?  Will it use string format?
Can you provide the sample yaml how it will look?

File: docs/proposals/mutex-improvements.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+* More sensible approach than including a new dependency (ConfigMap).
+* Relatively simple change.
+* Single source of truth
+
+##### Disadvantages of solution #1
+
+* Changing a field could be major change and might break projects in the ecosystem that rely on this behavior. This has been the behavior of mutexes ever since they have been implemented, it is high risk to change this in my opinion.
+
+#### Store holding/pending information in a config map
```
Comment (sarabala1979) [inline docs/proposals/mutex-improvements.md:69]: Can you add sample configmap content?

File: docs/proposals/mutex-improvements.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+We can store the holder keys inside a config map, on release, we refer to the holder keys inside this config map. There is an existing WIP PR for this, [here](https://github.com/argoproj/argo-workflows/pull/10009).
+It requires handling pending workflows. If we are going with this solution, a small amendment to deal with pending items will have to be made, we may have to introduce two config maps. One will be used for storing information regarding acquired locks, the other for storing information regarding pending lock acquisitions.
+There is a possibility that it might be possible to use a single config map here, but that solution needs to be explored in order to confirm this.
+
+##### Advantages of solution #2
+
+* The change in behavior is transparent to users.
+* Existing PR should allow for us to be quite quick in pushing a fix.
+
+##### Disadvantages of solution #2
```
Comment (sarabala1979) [inline docs/proposals/mutex-improvements.md:80]: how can we handle if holder information crosses the 1MB limit on configmap?

**Comment by JPZ13** [comment]
not stale, and not resolved

**Comment by Bwvolleyball** [comment]
This issue still feels relevant, right?

File: docs/proposals/mutex-improvements.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+`
+getHolderKey({"Namespace": "argo", Name: "example"}, "node") = "argo/example/node"
+
+-- MutexStatus After LockAcquired Call --
+items = ["argo", "example", "node"]
+holdingName = "node"
+ms.Holding = [MutexHolding{Mutex: lockKey, Holder: holdingName}]
+
+getResourceKey("argo", "example", "node") = "argo/example/node"
+`
```
Comment (agilgur5) [inline docs/proposals/mutex-improvements.md:23]:
```suggestion
\```go
getHolderKey({"Namespace": "argo", Name: "example"}, "node") = "argo/example/node"

// -- MutexStatus After LockAcquired Call --
items = ["argo", "example", "node"]
holdingName = "node"
ms.Holding = [MutexHolding{Mutex: lockKey, Holder: holdingName}]

getResourceKey("argo", "example", "node") = "argo/example/node"
\```
```

**Review by agilgur5 — COMMENTED** [review]
some formatting changes below that make the [rendered version](https://github.com/argoproj/argo-workflows/blob/49865b1783b481ba7600441999559821d1a03a18/docs/proposals/mutex-improvements.md) a good bit easier to read

File: docs/proposals/mutex-improvements.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+##### Advantages of solution #1
+
+* More sensible approach than including a new dependency (ConfigMap).
+* Relatively simple change.
+* Single source of truth
+
+##### Disadvantages of solution #1
+
+* Changing a field could be major change and might break projects in the ecosystem that rely on this behavior. This has been the behavior of mutexes ever since they have been implemented, it is high risk to change this in my opinion.
```
Comment (agilgur5) [inline docs/proposals/mutex-improvements.md:67]: Does this change a field on the CR itself? Or just the internal structs?

If it is just internal structs, a restart of the controller should get everything into the new format.

If it is on the CR, perhaps there could be a in-between option -- a struct that is encoded into a string?

Comment (isubasinghe) [reply docs/proposals/mutex-improvements.md:67]: I don't think it is just internal structs unfortunately, it gets serialised into json and protobufs hence why I assumed there might be some usage here from the outside world. 

Look, honestly seems very unlikely though, think we can probably just change the struct. Chucking everything in a ConfigMap would keep things consistent between different versions. 

I am open to just changing the struct. This is basically an RFC to get some idea of what to do.

Comment (agilgur5) [reply docs/proposals/mutex-improvements.md:67]: So I believe this impacts the `status.synchronization` field on the CR. We technically can't change the CRD without a breaking version change (e.g. `v1alpha1` -> `v1alpha2` or similar). 

We can _add_ fields though, as we do quite often. So perhaps we deprecate the existing ones and only use the new ones moving forward. In either manner, we would need to ensure backward-compat with older Workflows that used the old format.

I think that would mitigate the breaking change part and not have the drawbacks of additional ConfigMaps

**Comment by isubasinghe** [comment]
closed thanks to https://github.com/argoproj/argo-workflows/pull/13553
